# RNA-seq 分析流程

本流程为标准RNAseq分析流程，包括原始数据预处理、比对、融合基因检测、HLA定型等分析步骤。流程所需软件均已经打包至一个docker镜像。
在XDP上，通过docker-runner应用可以直接启动该镜像，通过一条命令行进行批量运行分析。该流程支持自动断点需跑，运行状态可视化。

## Table of Contents

- [Summary](#Summary)
- [Install](#install)
- [Tools](#tools)
- [Usage](#usage)
- [To do list](#todolist)
- [Maintainers](#maintainers)
- [License](#license)

## Summary

The workflow has the following main steps:
* Raw data processing using fastp
* Alignment to reference genome sequencing using STAR or sentieon STAR
* Quality control analysis using "Picard CollectRnaSeqMetrics"
* Gene/Transcript expression quantification using "Salmon" with alignment bam as input.
* Gene fusion analysis using STAR-Fusion
* HLA typing with the arcasHLA software

## Install

This workflow is developed to be run on [XDP](https://platform.xdp.basebit.me/user/login) without any installation. But it is also can be run locally with needed tools installed.

## Tools
* fastp:0.23.0
* (sentieon) STAR:2.7.8a
* STAR-Fusion:1.10.0
* picard: 2.20.3-SNAPSHOT
* salmon: 1.5.2
* arcasHLA:0.2.5

## Usage

### using docker runner mode
1. Log in XDP using Google Chrome browser.
2. Go to your project workspace and launch a docker instance with 'rnaseq_envs:1.2' image. remember to set enough attach_storage.
For computing resource, we encourage you to use spot instance to save money since our workflow support continuing run mode,
and you can always get back to your workspace as long as you keep the attach storage un-detached.
3. Then, change to "/root/attach_storage" directory. Run the following command:
```
python */rnaseq.py ....
```
4. Try to run the workflow again with above same command after you have solved the potential problems. It is will init the workflow at failed steps automatically.

### using APP mode
developing ......


### Arguments of workflow
```angular2html
optional arguments:
  -h, --help            show this help message and exit
  -star_index STAR_INDEX
                        star alignment index dir (default: None)
  --sentieon            indicate to use sentieon STAR tool (default: False)
  -fusion_index FUSION_INDEX
                        star-fusion database dir: CTAT_resource_lib from
                        https://data.broadinstitute.org/Trinity/CTAT_RESOURCE_LIB/__genome_libs_StarFv1.10/GRCh37_gencode_v19_CTAT_lib_Mar012021.plug-n-play.tar.gz
                        (default: None)
  -transcripts_fa TRANSCRIPTS_FA
                        transcriptome fasta file (default: None)
  -genome_fa GENOME_FA  genome fasta file, needed for variant calling (default: None)
  -gtf GTF              genome annotation file (default: None)
  -ref_flat REF_FLAT    gene model file, genome annotation file, download site: http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/refFlat.txt.gz (default: None)
  -rRNA_interval RRNA_INTERVAL
                        rRNA interval file generated by using picard BedToIntervalList command (default: None)
  -hla_db HLA_DB        arcasHLA database, please refer to https://github.com/RabadanLab/arcasHLA (default: None)
  -fastq_info FASTQ_INFO [FASTQ_INFO ...]
                        A list with elements from [fastq file, fastq parent dir, fastq_info.txt, fastq_info.json] (default: None)
  -r1_name R1_NAME      python regExp that describes the full name of read1 fastq file name. It requires at least one pair small brackets, and the string matched in the
                        first pair brackets will be used as sample name. Example: '(.*).R1.fq.gz' (default: (.*).R1.fastq)
  -r2_name R2_NAME      python regExp that describes the full name of read2 fastq file name. It requires at least one pair small brackets, and the string matched in the
                        first pair brackets will be used as sample name. Example: '(.*).R2.fq.gz' (default: (.*).R2.fastq)
  -exclude_samples EXCLUDE_SAMPLES [EXCLUDE_SAMPLES ...]
                        samples to exclude from analysis (default: ())
  -dbsnp DBSNP          dbsnp vcf file, only needed for variant calling (default: None)
  -known_indels KNOWN_INDELS
                        high confidence known indel vcf file, only needed for variant calling (default: None)
  -known_mills KNOWN_MILLS
                        high confidence known indel vcf file, only needed for variant calling (default: None)

Arguments for controlling running mode:
  --run                 运行流程，默认不运行, 仅生成流程，如果outdir目录已经存在cmd_state.txt文件，则自动续跑 (default: False)
  --docker              default won't use docker even if docker image is provided. (default: False)
  --plot                generate directed acyclic graph for whole workflow timely (default: False)
  -update_args update-args
                        输入参数配置文件, 其包含流程所有软件需要的参数，该配置文件设置的参数值将是流程中最后实际用的值 (default: None)
  -threads max-workers  允许的最大并行的cmd数目, 默认5 (default: 5)
  -outdir workdir       分析目录或结果目录 (default: /disk/biodatabase/tpipelines/Result)
  -skip step1 [task3 ...]
                        指定要跳过的步骤或具体task,空格分隔,默认程序会自动跳过依赖他们的步骤, 使用--list_cmd or --list_task可查看候选 (default: [])
  --no_skip_depend      当使用skip参数时, 如果同时指定该参数，则不会自动跳过依赖的步骤 (default: False)
  -rerun_steps task3 [task_prefix ...]
                        指定需要重跑的步骤，不论其是否已经成功完成，空格分隔, 这样做的可能原因可以是: 你重新设置了参数. 使用--list_task可查看候选，也可以使用task的前缀指定属于同一个步骤的task (default: [])
  -assume_success_steps task3 [task_prefix ...]
                        假定哪些步骤已经成功运行，不论其是否真的已经成功完成，空格分隔, 这样做的可能原因: 利用之前已经成功运行的结果. 使用--list_task可查看候选，也可以使用task的前缀指定属于同一个步骤的task (default: [])
  -retry max-retry      某步骤运行失败后再尝试运行的次数, 默认1次. 如需对某一步设置不同的值, 可在运行流程前修改pipeline.ini (default: 1)
  --list_cmd            仅仅显示当前流程包含的主步骤, 且已经排除指定跳过的步骤 (default: False)
  -show_cmd cmd-query   提供一个cmd名称,输出该cmd的样例 (default: None)
  --list_task           仅仅显示当前流程包含的详细步骤, 且已经排除指定跳过的步骤 (default: False)
  --monitor_resource    是否监控每一步运行时的资源消耗, 如需对某一步设置不同的值, 可在运行流程前修改pipeline.ini (default: False)
  -wait_resource_time wait-time
                        等待资源的时间上限, 默认每次等待时间为15秒, 等待时间超过这个时间且资源不足时判定任务失败 (default: 60)
  --no_check_resource_before_run
                        指示运行某步骤前检测指定的资源是否足够, 如不足, 则该步骤失败; 如果设置该参数, 则运行前不检查资源. 如需对某一步设置不同的值,可运行前修改pipeline.ini. 如需更改指定的资源, 可在运行流程前修改pipeline.ini (default: False)

```

### Create STAR index
Example cmd:
```angular2html
STAR --runThreadN 5 \
    --runMode genomeGenerate \
    --genomeDir star_index \
    --genomeFastaFiles chr22_with_ERCC92.fa \
    --sjdbGTFfile chr22_with_ERCC92_tidy.gtf \
    --sjdbOverhang 99 # read length -1
```


## To do list
- DESeq2
- Deconvolution of Immune Cell Proportions
- TCR analysis
- circRNA detection
- add StringTie for transcript assembling

## Maintainers

Email: *danny.gu@basebit.ai*

## License

[MIT © Richard McRichface.](../LICENSE)